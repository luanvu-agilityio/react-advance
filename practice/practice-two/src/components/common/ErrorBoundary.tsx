import React, { Component, type ReactNode } from 'react'
import styled from 'styled-components'
import { Theme, Button } from '@radix-ui/themes'
import { ShieldAlert } from 'lucide-react'
import Text from './Text'

interface Props {
  children: ReactNode
  fallback?: ReactNode
}

interface State {
  hasError: boolean
  error?: Error
}

const ErrorContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 400px;
  gap: 16px;
  padding: 2rem;
  text-align: center;
`

class ErrorBoundary extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError(error: Error): State {
    return { hasError: true, error }
  }

  componentDidCatch(error: Error, errorInfo: React.ErrorInfo) {
    console.error('Error caught by boundary:', error, errorInfo)
  }

  handleRetry = () => {
    this.setState({ hasError: false, error: undefined })
  }

  render() {
    if (this.state.hasError) {
      if (this.props.fallback) {
        return this.props.fallback
      }

      return (
        <ErrorContainer>
          <Theme>
            <ShieldAlert size={10} />
            <Text
              as="h2"
              text="Something went wrong"
              style={{
                fontSize: '24px',
                fontWeight: 'var(--font-weight-semibold)',
                color: 'var(--black-color-default)',
                margin: 0,
              }}
            />
            <Text
              as="p"
              text="Failed to load the page. This might be due to a network issue or temporary problem."
              style={{
                fontSize: '16px',
                color: 'var(--black-shade-2)',
                margin: 0,
                maxWidth: '500px',
              }}
            />
            <Button
              variant="solid"
              onClick={this.handleRetry}
              style={{
                backgroundColor: 'var(--green-color-default)',
                color: 'white',
              }}
            >
              Try Again
            </Button>
          </Theme>
        </ErrorContainer>
      )
    }

    return this.props.children
  }
}

export default ErrorBoundary
