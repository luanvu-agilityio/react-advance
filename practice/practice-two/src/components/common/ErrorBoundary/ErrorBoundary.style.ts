import styled from 'styled-components'
import { Button } from '@radix-ui/themes'
import { ShieldAlert } from 'lucide-react'

export const ErrorContainer = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  min-height: 400px;
  gap: 16px;
  padding: 2rem;
  text-align: center;
`

export const ThemeWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 16px;
`

export const StyledShieldIcon = styled(ShieldAlert)`
  color: var(--orange-color-default);
  margin-bottom: 8px;
`

export const ErrorTitle = styled.h2`
  font-size: 24px;
  font-weight: var(--font-weight-semibold);
  color: var(--black-color-default);
  margin: 0;
  text-align: center;
`

export const ErrorDescription = styled.p`
  font-size: 16px;
  color: var(--black-shade-2);
  margin: 0;
  max-width: 500px;
  text-align: center;
  line-height: 1.5;
`

export const RetryButton = styled(Button)`
  background-color: var(--green-color-default);
  color: white;
  padding: 12px 24px;
  border-radius: 8px;
  font-size: 14px;
  font-weight: var(--font-weight-semibold);

  &:hover {
    background-color: var(--green-shade-1);
    transform: translateY(-1px);
    transition: all 0.2s ease;
  }

  &:focus {
    box-shadow: 0 0 0 2px var(--green-shade-2);
    outline: none;
  }

  &:active {
    transform: translateY(0);
  }
`
