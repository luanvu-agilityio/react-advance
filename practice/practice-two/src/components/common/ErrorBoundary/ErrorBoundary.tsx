import { Component, type ErrorInfo, type ReactNode } from 'react'
import { Theme } from '@radix-ui/themes'
import {
  ErrorContainer,
  ThemeWrapper,
  StyledShieldIcon,
  ErrorTitle,
  ErrorDescription,
  RetryButton,
} from './ErrorBoundary.style'

export interface ErrorBoundaryProps {
  children: ReactNode
  fallback?: ReactNode
  onError?: (error: Error, errorInfo: ErrorInfo) => void
}

interface ErrorBoundaryState {
  hasError: boolean
  error?: Error
}

class ErrorBoundary extends Component<ErrorBoundaryProps, ErrorBoundaryState> {
  constructor(props: ErrorBoundaryProps) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError(error: Error): ErrorBoundaryState {
    return { hasError: true, error }
  }

  componentDidCatch(error: Error, errorInfo: ErrorInfo) {
    console.error('Error caught by boundary:', error, errorInfo)

    // Call optional error handler
    this.props.onError?.(error, errorInfo)
  }

  handleRetry = () => {
    this.setState({ hasError: false, error: undefined })
  }

  render() {
    if (this.state.hasError) {
      // Use custom fallback if provided
      if (this.props.fallback) {
        return this.props.fallback
      }

      // Default error UI
      return (
        <ErrorContainer>
          <Theme>
            <ThemeWrapper>
              <StyledShieldIcon size={48} />

              <ErrorTitle>Something went wrong</ErrorTitle>

              <ErrorDescription>
                Failed to load the page. This might be due to a network issue or
                temporary problem.
              </ErrorDescription>

              <RetryButton variant="solid" onClick={this.handleRetry}>
                Try Again
              </RetryButton>
            </ThemeWrapper>
          </Theme>
        </ErrorContainer>
      )
    }

    return this.props.children
  }
}
export default ErrorBoundary
