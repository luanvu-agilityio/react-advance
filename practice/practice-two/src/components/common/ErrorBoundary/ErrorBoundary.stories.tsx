import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'
import ErrorBoundary from './ErrorBoundary'

// Component that throws an error for testing
const ThrowError = ({ shouldThrow }: { shouldThrow: boolean }) => {
  if (shouldThrow) {
    throw new Error('Test error for ErrorBoundary')
  }
  return <div>This component works fine!</div>
}

const meta: Meta<typeof ErrorBoundary> = {
  title: 'Components/Common/ErrorBoundary',
  component: ErrorBoundary,
  parameters: {
    layout: 'fullscreen',
    docs: {
      description: {
        component:
          'ErrorBoundary catches JavaScript errors anywhere in the child component tree and displays a fallback UI instead of crashing the entire application.',
      },
    },
  },
  argTypes: {
    children: {
      description: 'Child components to wrap with error boundary',
    },
    fallback: {
      description: 'Custom fallback UI to display when an error occurs',
    },
    onError: {
      description: 'Callback function called when an error is caught',
      action: 'onError',
    },
  },
  tags: ['autodocs'],
}

export default meta
type Story = StoryObj<typeof ErrorBoundary>

export const Default: Story = {
  render: () => (
    <ErrorBoundary onError={action('error-caught')}>
      <ThrowError shouldThrow={true} />
    </ErrorBoundary>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Default error boundary showing the built-in error UI when a child component throws an error.',
      },
    },
  },
}

export const WorkingComponent: Story = {
  render: () => (
    <ErrorBoundary onError={action('error-caught')}>
      <ThrowError shouldThrow={false} />
    </ErrorBoundary>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Error boundary with a working child component - no error UI is shown.',
      },
    },
  },
}

export const CustomFallback: Story = {
  render: () => (
    <ErrorBoundary
      fallback={
        <div
          style={{
            padding: '2rem',
            textAlign: 'center',
            backgroundColor: '#fff3cd',
            border: '1px solid #ffeaa7',
            borderRadius: '8px',
          }}
        >
          <h3>Custom Error Message</h3>
          <p>This is a custom fallback UI</p>
        </div>
      }
      onError={action('error-caught')}
    >
      <ThrowError shouldThrow={true} />
    </ErrorBoundary>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Error boundary with a custom fallback UI instead of the default error display.',
      },
    },
  },
}

export const NestedComponents: Story = {
  render: () => (
    <ErrorBoundary onError={action('error-caught')}>
      <div style={{ padding: '20px' }}>
        <h2>Parent Component</h2>
        <div style={{ margin: '20px 0' }}>
          <h3>Child Component</h3>
          <ThrowError shouldThrow={true} />
        </div>
      </div>
    </ErrorBoundary>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'Error boundary wrapping nested components where a deeply nested component throws an error.',
      },
    },
  },
}
