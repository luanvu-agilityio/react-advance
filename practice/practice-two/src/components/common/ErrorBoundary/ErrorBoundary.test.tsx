import { render, screen, fireEvent } from '@testing-library/react'
import ErrorBoundary from './ErrorBoundary'
import type { ReactNode } from 'react'

// Mock the styled components
jest.mock('./ErrorBoundary.style', () => ({
  ErrorContainer: ({ children }: { children: ReactNode }) => (
    <div data-testid="error-container">{children}</div>
  ),
  ThemeWrapper: ({ children }: { children: ReactNode }) => (
    <div data-testid="theme-wrapper">{children}</div>
  ),
  StyledShieldIcon: ({ size }: { size: number }) => (
    <div data-testid="shield-icon" data-size={size}>
      ShieldIcon
    </div>
  ),
  ErrorTitle: ({ children }: { children: ReactNode }) => (
    <h2 data-testid="error-title">{children}</h2>
  ),
  ErrorDescription: ({ children }: { children: ReactNode }) => (
    <p data-testid="error-description">{children}</p>
  ),
  RetryButton: ({
    children,
    onClick,
    variant,
  }: {
    children: ReactNode
    onClick: () => void
    variant?: string
  }) => (
    <button data-testid="retry-button" onClick={onClick} data-variant={variant}>
      {children}
    </button>
  ),
}))

// Mock Radix UI Theme
jest.mock('@radix-ui/themes', () => ({
  Theme: ({ children }: { children: ReactNode }) => (
    <div data-testid="radix-theme">{children}</div>
  ),
}))

// Component that throws an error for testing
const ThrowError = ({ shouldThrow }: { shouldThrow: boolean }) => {
  if (shouldThrow) {
    throw new Error('Test error')
  }
  return <div data-testid="working-component">Working component</div>
}

// Suppress console.error for tests
const originalError = console.error
beforeAll(() => {
  console.error = jest.fn()
})

afterAll(() => {
  console.error = originalError
})

describe('ErrorBoundary', () => {
  // SNAPSHOT TESTS FOR RENDERING STATES
  describe('Rendering States', () => {
    it('renders correctly when there is no error', () => {
      const { container } = render(
        <ErrorBoundary>
          <ThrowError shouldThrow={false} />
        </ErrorBoundary>
      )
      expect(container).toMatchSnapshot('no-error')
    })

    it('renders error UI correctly when error occurs', () => {
      const { container } = render(
        <ErrorBoundary>
          <ThrowError shouldThrow={true} />
        </ErrorBoundary>
      )
      expect(container).toMatchSnapshot('with-error')
    })

    it('renders custom fallback correctly', () => {
      const customFallback = (
        <div data-testid="custom-fallback">Custom Error UI</div>
      )
      const { container } = render(
        <ErrorBoundary fallback={customFallback}>
          <ThrowError shouldThrow={true} />
        </ErrorBoundary>
      )
      expect(container).toMatchSnapshot('custom-fallback')
    })
  })

  // INTERACTIVE BEHAVIOR TESTS
  describe('Interactive Behavior', () => {
    it('retries when retry button is clicked', () => {
      let shouldThrow = true

      const TestComponent = () => {
        if (shouldThrow) {
          throw new Error('Test error')
        }
        return <div data-testid="recovered-component">Component recovered</div>
      }

      render(
        <ErrorBoundary>
          <TestComponent />
        </ErrorBoundary>
      )

      // Error UI should be shown
      expect(screen.getByTestId('error-container')).toBeInTheDocument()

      // Prepare for retry - this will make the component render without error next time
      shouldThrow = false

      // Click retry button
      fireEvent.click(screen.getByTestId('retry-button'))

      // Component should recover
      expect(screen.getByTestId('recovered-component')).toBeInTheDocument()
      expect(screen.queryByTestId('error-container')).not.toBeInTheDocument()
    })

    it('calls onError callback when error occurs', () => {
      const onError = jest.fn()

      render(
        <ErrorBoundary onError={onError}>
          <ThrowError shouldThrow={true} />
        </ErrorBoundary>
      )

      // Verify onError was called with correct parameters
      expect(onError).toHaveBeenCalledWith(
        expect.any(Error),
        expect.objectContaining({
          componentStack: expect.any(String),
        })
      )
    })
  })
})
