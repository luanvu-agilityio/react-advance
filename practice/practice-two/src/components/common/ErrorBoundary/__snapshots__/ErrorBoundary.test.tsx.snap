// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary Rendering States renders correctly when there is no error: no-error 1`] = `
<div>
  <div
    data-testid="working-component"
  >
    Working component
  </div>
</div>
`;

exports[`ErrorBoundary Rendering States renders custom fallback correctly: custom-fallback 1`] = `
<div>
  <div
    data-testid="custom-fallback"
  >
    Custom Error UI
  </div>
</div>
`;

exports[`ErrorBoundary Rendering States renders error UI correctly when error occurs: with-error 1`] = `
<div>
  <div
    data-testid="error-container"
  >
    <div
      data-size="48"
      data-testid="shield-icon"
    >
      ShieldIcon
    </div>
    <h2
      data-testid="error-title"
    >
      Something went wrong
    </h2>
    <p
      data-testid="error-description"
    >
      Failed to load the page. This might be due to a network issue or temporary problem.
    </p>
    <button
      data-testid="retry-button"
      data-variant="solid"
    >
      Try Again
    </button>
  </div>
</div>
`;
